### PR TITLE
Add welt.de extractor.

### LINKED ISSUE #17553: welt.de - always downloads the worst quality.
### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

https://www.welt.de/mediathek/dokumentation/natur-und-wildlife/sendung180126234/Wild-Japan-Land-der-tausend-Inseln-Honshu.html
youtube-dl can't get to m3u8 playlist, and goes straight to last video named "*_200.mp4", which is worst of all, "*_2000.mp4"- is the best.

`<video id="video-02f5d6ad-9e72-4924-a3ec-275354e7a422" class="u-display--is-hidden" width="auto" height="0" preload="none" data-qa="Video.id"> <source data-content="Video.source" src="https://weltmediathek-vh.akamaihd.net/i/2018/07/12577130_12577134_PG0009J1/12577130_12577134_PG0009J1_,2000,1500,1000,200,.mp4.csmil/master.m3u8" type="application/x-mpegURL" data-mediatype="hls" data-default="true" data-qa="Video.source"/> <source data-content="Video.source" src="https://weltn24lfthumb-a.akamaihd.net/2018/07/12577130_12577134_PG0009J1/12577130_12577134_PG0009J1_2000.mp4" type="video/mp4" data-mediatype="mp4" data-default="false" data-qa="Video.source"/> <source data-content="Video.source" src="https://weltn24lfthumb-a.akamaihd.net/2018/07/12577130_12577134_PG0009J1/12577130_12577134_PG0009J1_1500.mp4" type="video/mp4" data-mediatype="mp4" data-default="false" data-qa="Video.source"/> <source data-content="Video.source" src="https://weltn24lfthumb-a.akamaihd.net/2018/07/12577130_12577134_PG0009J1/12577130_12577134_PG0009J1_1000.mp4" type="video/mp4" data-mediatype="mp4" data-default="false" data-qa="Video.source"/> <source data-content="Video.source" src="https://weltn24lfthumb-a.akamaihd.net/2018/07/12577130_12577134_PG0009J1/12577130_12577134_PG0009J1_200.mp4" type="video/mp4" data-mediatype="mp4" data-default="false" data-qa="Video.source"/> </video> `

Additional urls:
https://www.welt.de/mediathek/dokumentation/natur-und-wildlife/sendung155954102/In-ewiger-Dunkelheit-Geschoepfe-der-Tiefsee.html
https://www.welt.de/mediathek/dokumentation/technik-und-wissen/sendung169520775/Boeing-787-Ein-Dreamliner-wird-ausgeliefert.html

### PR DESCRIPTION
**Pull request by MPW1412**
Hello,

this is my first PR for youtube-dl. So please just let me know, if I don't match the standards here.

Question: How could I provide multiple formats? The videos on welt.de are usually provided in four different resolutions, which can be determined by the _xxx.mp4 in the URL. 

Fixes #17553

- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### purpose of this *pull request*?
- [x] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

Basic function for welt.de. Instead of loading the worst quality with the generic extractor, you'll get the best quality.

TODO: Provide list of all qualities and resolutions available.

### CODE REVIEW

**Comment by MPW1412** [comment]
@dstftw, is there a reason you don't merge this one? Please let me know if the code needs improvement or how I could include all qualities available.

**Comment by dstftw** [comment]
1. Read tutorial and coding conventions.
2. All formats must be extracted and sorted appropriately.

**Comment by MPW1412** [comment]
@dstftw, @remitamine, I added another commit. All formats are available for download now. I hope this can be merged now.

**Comment by MPW1412** [comment]
@dstftw, @remitamine, any update on this?

**Comment by MPW1412** [comment]
Yet another month passed. I think I matched all coding conventions. Please merge this or let me know, what's wrong with the code.

File: youtube_dl/extractor/welt.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    ]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        title = remove_end(self._html_search_regex(r'<title>([^<]+)</title>', webpage, 'title')
+                           .strip(), ' - Video - WELT')
+
+        sources = self._search_regex(r'(["\'])?sources\1\s*:\s*\[({\s*\1file\1\s*:\s*\1([^\1,])+\1\s*}\s*,?)+', webpage, 'sources', group=0)
+        files = re.findall(r'http[^\'"]+', sources)
```
Comment (dstftw) [inline youtube_dl/extractor/welt.py]: `self._parse_html5_media_entries` for formats extraction.

File: youtube_dl/extractor/welt.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        formats = []
+        for url in files:
+            number = re.search(r'_(\d+)\.mp4', url).group(1)
+            formats.append({
+                'url': url,
+                'quality_key': int_or_none(number)
+            })
+        self._remove_duplicate_formats(formats)
+        formats = sorted(formats, key=lambda x: x['quality_key'])
```
Comment (dstftw) [inline youtube_dl/extractor/welt.py]: `_sort_formats`.

File: youtube_dl/extractor/welt.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        webpage = self._download_webpage(url, video_id)
+        title = remove_end(self._html_search_regex(r'<title>([^<]+)</title>', webpage, 'title')
+                           .strip(), ' - Video - WELT')
+
+        sources = self._search_regex(r'(["\'])?sources\1\s*:\s*\[({\s*\1file\1\s*:\s*\1([^\1,])+\1\s*}\s*,?)+', webpage, 'sources', group=0)
+        files = re.findall(r'http[^\'"]+', sources)
+
+        formats = []
+        for url in files:
+            number = re.search(r'_(\d+)\.mp4', url).group(1)
```
Comment (dstftw) [inline youtube_dl/extractor/welt.py]: Bitrate should go to corresponding format meta field.

File: youtube_dl/extractor/welt.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        self._remove_duplicate_formats(formats)
+        formats = sorted(formats, key=lambda x: x['quality_key'])
+        quality_counter = -1
+        for i in range(len(formats) - 1, 0, -1):
+            formats[i] = {'url': formats[i]['url'], 'quality': quality_counter}
+            quality_counter -= 1
+
+        return {
+            'id': video_id,
+            'ext': 'mp4',
```
Comment (dstftw) [inline youtube_dl/extractor/welt.py]: Not used with `formats`.

File: youtube_dl/extractor/welt.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+from ..utils import (
+    remove_end,
+    int_or_none
+)
+
+
+class WeltIE(InfoExtractor):
+    IE_NAME = 'welt.de'
+    _VALID_URL = r'''https?://(?:www\.)?welt\.de/[^\.]+/(?P<id>[a-z]+\d+)(?:/.*)?'''
```
Comment (dstftw) [inline youtube_dl/extractor/welt.py]: Optional group at the end does not make any sense. Triple quotes are pointless.

File: youtube_dl/extractor/welt.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
+from __future__ import unicode_literals
+
+import re
+
+from .common import InfoExtractor
+
+from ..utils import (
+    remove_end,
+    int_or_none
```
Comment (dstftw) [inline youtube_dl/extractor/welt.py]: Alphabetic.

**Pull request closed without merging**